### PR TITLE
Fix missed cutout data when building renewable profile

### LINKED ISSUE #510: TopologicalError in build_renewable_profile when cutout has some missed data
An ERA5-derived cutout sometimes may contain missed data. The reason is some technical issues related to availability on Copernicus Climate Data store as discussed in [this forum thread](https://confluence.ecmwf.int/display/CUSF/Missing+data+in+ERA5T). Such data loss is visible on the cutout parameters fields as white stripes (@davide-f, many thanks for reporting!), e.g.:

```python
cutout = atlite.Cutout(cutout_path)
cutout.data["wnd100m"][5].plot()
```

cutout_data_loss

When being present into cutout data, this data flaw leads to the TopologicalError problem in `build_renewable_profiles`. A full error listing looks like that:

```bash
INFO:__main__:Calculate underwater fraction of connections.
ERROR:shapely.geos:IllegalArgumentException: CGAlgorithmsDD::orientationIndex encountered NaN/Inf numbers
INFO:shapely.geos:Invalid Coordinate at or near point nan nan
Traceback (most recent call last):
  File "~/Documents/_github_/pypsa-earth/.snakemake/scripts/tmpfsp8o7r7.build_renewable_profiles.py", line 586, in <module>
    frac = line.intersection(offshore_shape).length / line.length
  File "~/opt/miniconda3/envs/pypsa-earth/lib/python3.10/site-packages/shapely/geometry/base.py", line 695, in intersection
    return geom_factory(self.impl['intersection'](self, other))
  File "~/opt/miniconda3/envs/pypsa-earth/lib/python3.10/site-packages/shapely/topology.py", line 73, in __call__
    self._check_topology(err, this, other)
  File "~/opt/miniconda3/envs/pypsa-earth/lib/python3.10/site-packages/shapely/topology.py", line 38, in _check_topology
    raise TopologicalError(
shapely.errors.TopologicalError: The operation 'GEOSIntersection_r' could not be performed. Likely cause is invalidity of the geometry <shapely.geometry.linestring.LineString object at 0x149b28a30>
[Sun Nov  6 20:21:35 2022]
INFO:snakemake.logging:[Sun Nov  6 20:21:35 2022]
Error in rule build_renewable_profiles:
    jobid: 16
    output: resources/renewable_profiles/profile_offwind-ac.nc
    log: logs/build_renewable_profile_offwind-ac.log (check log file(s) for error message)

ERROR:snakemake.logging:Error in rule build_renewable_profiles:
    jobid: 16
    output: resources/renewable_profiles/profile_offwind-ac.nc
    log: logs/build_renewable_profile_offwind-ac.log (check log file(s) for error message)

RuleException:
CalledProcessError in line 355 of ~/Documents/_github_/pypsa-earth/Snakefile:
Command 'set -euo pipefail;  ~/opt/miniconda3/envs/pypsa-earth/bin/python3.10 ~/Documents/_github_/pypsa-earth/.snakemake/scripts/tmpfsp8o7r7.build_renewable_profiles.py' returned non-zero exit status 1.
  File "~/Documents/_github_/pypsa-earth/Snakefile", line 355, in __rule_build_renewable_profiles
  File "~/opt/miniconda3/envs/pypsa-earth/lib/python3.10/concurrent/futures/thread.py", line 58, in run
ERROR:snakemake.logging:RuleException:
CalledProcessError in line 355 of ~/Documents/_github_/pypsa-earth/Snakefile:
Command 'set -euo pipefail;  ~/opt/miniconda3/envs/pypsa-earth/bin/python3.10 ~/Documents/_github_/pypsa-earth/.snakemake/scripts/tmpfsp8o7r7.build_renewable_profiles.py' returned non-zero exit status 1.
  File "~/Documents/_github_/pypsa-earth/Snakefile", line 355, in __rule_build_renewable_profiles
  File "~/opt/miniconda3/envs/pypsa-earth/lib/python3.10/concurrent/futures/thread.py", line 58, in run
Shutting down, this might take some time.
WARNING:snakemake.logging:Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
ERROR:snakemake.logging:Exiting because a job execution failed. Look above for error message
Complete log: .snakemake/log/2022-11-06T201853.281672.snakemake.log
WARNING:snakemake.logging:Complete log: .snakemake/log/2022-11-06T201853.281672.snakemake.log
```

### PR DESCRIPTION
**Pull request by ekatef**
# Closes #510 

## Changes proposed in this Pull Request
Deal with situations with empty cutouts without breaking the code, e.g. zero capacity factor when calculating RES-related geometries, AND inform by new functions about missing data/ information.
 
## Checklist

- [x] I tested my contribution locally and it seems to work fine.
    - [x] on a cutout with some data loss
    - [x] on a full cutout
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [x] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by pz-max** [comment]
@ekatef let's discuss that in the next weekly meeting. I am adding a note.
I would recommend making the test/ check a separate function.
This would be easier to maintain as it has a clear purpose with an e.g. docstring

**Comment by ekatef** [comment]
@pz-max, you mean to put all the GIS-checking functions into a separate script? I think it could perfectly make sense. The only problem I see here may be de-coupling of the testing functions from local calculations

**Comment by pz-max** [comment]
@ekatef I think we could start making the test a function in the same script. If we have some of them we could put that in a GIS test script

**Comment by ekatef** [comment]
@pz-max, it'll be interesting to discuss live. Thanks for adding the note :)

**Comment by davide-f** [comment]
Is this PR ready?
It looks good to me

**Comment by ekatef** [comment]
I think yes. But I would re-run tests on the fresh configuration

**Comment by ekatef** [comment]
Update: the PR have been tested on the fresh project configuration and worked fine.

@davide-f, I think it's ready for the review

**Comment by davide-f** [comment]
For me looks good, @pz-max had comments instead.
Let's align on that.
If no answer, I'd squash and merge in the coming days, many thanks!

**Comment by pz-max** [comment]
@davide-f @ekatef , yes I would indeed do smt different. Just looking at my face in a couple of months assessing the code, it won't look nice! I recommend to write the check therefore in a function with a docstring and only call the function at the certain location

Example:
```python
def cutout_test(capacity_factor):
        """
        Check if a cutout contains zero's aka. misses values

        CDS-extracted ERA5 data sometimes may contain missed values
        https://confluence.ecmwf.int/display/CUSF/Missing+data+in+ERA5T
        """
        n_missed_cells = pd.isnull(capacity_factor).sum()
        n_cells = len(np.ndarray.flatten(capacity_factor.data))
        share_missed_cells = 100 * (n_missed_cells / n_cells)
        if share_missed_cells > 0:
            logger.warning(
                f"The provided cutout contains missed data:\r\n content of {share_missed_cells:2.1f}% all cutout cells is lost"
            )
       if share_missed_cells > 0.05:
              raise ValueError(f"The cutout misses more than 5% of data, in total {share_missed_cells:2.1f}% are missed.")
```

File: scripts/build_renewable_profiles.py [hunk 1]
```diff
@@ -532,6 +543,9 @@ def get_hydro_capacity_annual_hydro_generation(
             ]
         )
 
+        n_weights_initial = len(ds.weight)
+        n_lost_weights = pd.isnull(ds.weight).sum()
```
Comment (pz-max) [inline scripts/build_renewable_profiles.py]: Can we move that to line 572?

File: scripts/build_renewable_profiles.py [hunk 2]
```diff
@@ -481,6 +483,15 @@ def get_hydro_capacity_annual_hydro_generation(
...
 
+        # CDS-extracted ERA5 data sometimes may contain missed values
+        # https://confluence.ecmwf.int/display/CUSF/Missing+data+in+ERA5T
+        n_missed_cells = pd.isnull(capacity_factor).sum()
+        n_cells = len(np.ndarray.flatten(capacity_factor.data))
+        share_missed_cells = 100 * (n_missed_cells / n_cells)
+        if share_missed_cells > 0:
+            logger.warning(
+                f"The provided cutout contains missed data:\r\n content of {share_missed_cells:2.1f}% all cutout cells is lost"
+            )
```
Comment (pz-max) [inline scripts/build_renewable_profiles.py]: Recommended to put into function

File: scripts/build_renewable_profiles.py [hunk 3]
```diff
@@ -556,4 +570,18 @@ def get_hydro_capacity_annual_hydro_generation(
...
+
+        if share_missed_weights2 >= 30:
+            recommend_msg = "\r\nYou may want to re-generate the cutout"
+        else:
+            recommend_msg = ""
+
+        if n_lost_weights > 0:
+            logger.warning(
+                f"Missed cutout data have resulted in data loss:\r\n for {tech} {share_missed_weights:2.1f}% buses overall and {share_missed_weights2:2.1f}% of buses with non-zero capacities {recommend_msg}"
+            )
```
Comment (pz-max) [inline scripts/build_renewable_profiles.py]: Recommended to put into a function

**Review by pz-max — CHANGES_REQUESTED** [review]

File: scripts/build_renewable_profiles.py (hunk 1 above)
Comment (ekatef) [reply scripts/build_renewable_profiles.py]: Yes, and we can also simplify a bit a test on a share of lost buses

File: scripts/build_renewable_profiles.py (hunk 2 above)
Comment (ekatef) [reply scripts/build_renewable_profiles.py]: Done

File: scripts/build_renewable_profiles.py (hunk 3 above)
Comment (ekatef) [reply scripts/build_renewable_profiles.py]: Done along with simplifying a warning

**Comment by ekatef** [comment]
@davide-f, thank you for looking into it!

@pz-max, a good point regarding a couple of months :) Have done some refactoring and will be grateful for your review

**Review by pz-max — APPROVED** [review]

**Comment by pz-max** [comment]
@ekatef amazing contribution :1st_place_medal:

**Pull request merged**